docs/proposals/20210126-nc-and-ncp.md

```diff
@@ -0,0 +1,304 @@
+	// +optional
+	ControllerManagerRef *corev1.ObjectReference `json:"controllerManager,omitempty"`
+}
+
+type NestedControlPlaneStatus struct {
+	// Etcd stores the connection information from the downstream etcd 
+	// implementation if the NestedEtcd type isn't used this
+	// allows other component controllers to fetch the endpoints.
+	// +optional
+	Etcd *NestedControlPlaneStatusEtcd `json:"etcd,omitempty"`
+
+	// APIServer stores the connection information from the control plane
+	// this should contain anything shared between control plane components
+	// +optional
+	APIServer *NestedControlPlaneStatusAPIServer `json:"apiserver,omitempty"`
+    
+	// Initialized denotes whether or not the control plane has the
+	// uploaded kubernetes config-map.
+	// +optional
+	Initialized bool `json:"initialized"`
+    
+	// Ready denotes that the AWSManagedControlPlane API Server is ready to
+	// receive requests and that the VPC infra is ready.
+	// +kubebuilder:default=false
+	Ready bool `json:"ready"`
+    
+	// ErrorMessage indicates that there is a terminal problem reconciling the
+	// state, and will be set to a descriptive error message.
+	// +optional
+	FailureMessage *string `json:"failureMessage,omitempty"`
+    
+	// Conditions specifies the cpnditions for the managed control plane
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+type NestedControlPlaneStatusEtcd struct {
+	// Addresses defines how to address the etcd instance
+	Addresses []NestedEtcdAddress `json:"addresses,omitempty"`
+}
+
+type NestedControlPlaneStatusAPIServer struct {
+	// ServiceCIDRs which is provided to kube-apiserver and kube-controller-manager
+	// +optional
+	ServiceCIDR string `json:"serviceCidr,omitempty"`
+}
+```
+
+### NestedControlPlane Lifecycle
+
+![Creating a NestedControl Plane](images/nestedcontrolplane/ncp-activity.png)
+
+#### Component CustomResources (NEtcd, NKAS, NKCM)
+
+After the controller has fetched the `NestedControlPlane` resource and that it's not in a terminating state it will look for all the [Nested* Component Controllers](20201026-creating-control-plane-components.md). It will do this by following the `*Refs` defined on the `NestedControlPlaneSpec`. Once all of the Nested* component Custom Resources have been found the controller will patch these objects setting the `metadata.ownerReferences` back to the `NestedControlPlane`.
+
+```go
+if err := controllerutil.SetControllerReference(ncp, component, r.scheme); err != nil {
+    // Return errors
+}
+```
+
+
+### PKI/Certificate Management
+
+Within a CAPN Cluster we need to manage the certificates and CA's which are used for the components in the control plane, this MUST be done outside of each of the component controllers so that we can share this information. As well these should be overridable by supplying your own certificates using `cert-manager` or other solutions by the use of shared common Kubernetes secret names.
+
```

Comment:
Will the NCP controller generate the certificate key pair for each component? Or each nested controller will need to generate the cert key pair based on the CA created by the NCP controller?

Reply:
Following up on this based on the conversation we had in Slack, I'm going to update the doc to reflect using secret names for the cert/key pairs as well that the NCP will generate. This is a place where our implementation will differ from KCP/CAPBK